docs/way-of-working/guides/testing/testing.md

```diff
@@ -1 +1,132 @@
-## Testing overview
\ No newline at end of file
+---
+title: Way of Testing
+sidebar_position: 10
+---
+
+:::danger[Draft]
+
+This document is still in draft. You can discuss it as part of this PR: https://github.com/dhis2/wow-frontend/pull/7.
+
+The scope of this document is automated tests _in code_; unit, integration, end to end, contract etc... in the context of the _frontend_ for platform team. It does not tackle other important aspects such as manual testing and its processes, or backend and database tests.
+:::
+
+### Why automated tests?
+
+Why we write tests? Good tests can...
+
+1. Verify the code is working correctly...
+2. Prevent future regressions
+3. Document code behaviour
+4. Provides design guidance
+5. Support refactoring
+
+> Read the full article here:  https://madeintandem.com/blog/five-factor-testing/
+
+
+:::warning[Discussion]
+
+- Rank these reasons in terms of importance in our context according to you
+- Any other reasons we should write tests?
+
+:::
+
+
+### The ideal scenario
+
+(Mozafar): we have a **high coverage** of tests, _integration-like_ with the most crucial scenarios tested in a thin layer of _end to end_ tests. Regressions and breaking changes in the API are caught early by tests, and refactoring made easy.
+
+:::warning[Discussion]
+
+- How does the ideal scenario look for you? some other possible prompts: TDD, cucumber, contract testing, end to end, pair programming... 
+
+:::
+
+:::warning[Discussion]
+
+- What should be a high coverage in our context?
+- What is unit vs integration tests in the frontend?
```

Comment:
See [my comment above](https://github.com/dhis2/wow-frontend/pull/7/files#r1671740093).
I think we shouldn't care about the definitions and rather talk about what checkboxes code must tick so we add tests.

Reply:
I agree that code examples on what we want to tests and how are  can be a great way to start, rather than definitions. What do you think about postponing the definition here and add instead a link to some good tests example in the code bases, which we think could be used as a general basis? If we start with there, it would give the team a good place to start writing more of each of the tests. Then as we write more tests, and we improve and get a better understanding, we can of think about the definitions and aim for continuous improvements until we are happy with it

Reply:
Sure can do!

_ASIDE: There's something I'd like to talk about before going through an example: IMO, it's impossible to have a rule that tells you exactly which files to test. In the end, it's the developer who has to judge whether a file / functionality falls under a certain category or not. Another reason why it's impossible is that sometimes it's hard to test something specific and easier to test its internally used components (I'll give an example for that). So in the end, it's up to the developer to decide what's worth testing and with which tools._

I think the [aggregate data entry app](https://github.com/dhis2/aggregate-data-entry-app/tree/master)'s source code is a very good example for several reasons:

* It's architecture is consistent
* Code is divided into "abstractions" (the `src/shared` folder) and implementation (all the other folders)
* This app serves as a good example of having too many Cypress tests instead of Jest tests
* This app uses `index.js`-files consistently and for re-exporting only

When you look inside the `src/shared` folder, that's what I'd call cohesive generalizations:

* They cover a specific concept
* They work in isolation (they're not written in a way to cater to a specific other module, folder, or concept)

Then there are other modules that use those generalizations to construct a very specific/concrete component:

* They're not meant for re-use
* They use the generalizations + module-internal code

Because this "app uses `index.js`-files consistently and for re-exporting only", we can take a look at what's being exported from those modules. I think the `src/context-selection` folder is a good place to start looking at:

* This is an example of "_sometimes it's hard to test something specific and easier to test its internally used components_"
* It's still relatively easy to understand what should be tested.

The `src/context-selection/index.js` has two exports: [`ContextSelection`](https://github.com/dhis2/aggregate-data-entry-app/blob/master/src/context-selection/context-selection/context-selection.js) and `ContextualHelpSidebar`. Testing `ContextualHelpSidebar` is fairly trivial. Testing the `ContextSelection` is not easy, though, simply because it's covering so many different topics.

This is where the developer has to make a choice, and the part which can't be covered by rules / automation:

1. The component can be tested with cypress (the choice I made; in hindsight, I think the other choice would've been better)
2. The `ContextSelection` can be unit-tested (mocking internally used components and just testing the actual "implementation" code), and then we can add tests for the other components/hooks used internally

The reason why I think this shouldn't have been done with Cypress is that the tests I've added don't cover critical user paths, but instead test the component in detail.

So when we look at the implementation of the [`ContextSelection`](https://github.com/dhis2/aggregate-data-entry-app/blob/master/src/context-selection/context-selection/context-selection.js), we can see that it's composing different components. None of the components are abstractions / generalizations, they're concretions!
So in this case, it makes sense to add tests for the following components:

* ContextualHelpSidebar (as it's exported by the `index.js`)
* ContextSelection (but in the form of a unit test, because we'll test)
* DataSetSelectorBarItem
* OrgUnitSetSelectorBarItem
* PeriodSelectorBarItem
* AttributeOptionComboSelectorBarItem
* SectionFilterSelectorBarItem

By having tests for these, we:

* cover the behavior that we care about directly
* cover the correctness of the `shared` generalizations indirectly
* cover the module contracts (their interfaces/ boundaries), as that's essentially what the `index.js` are (the contract of how external modules use the module in question and what information is transferred during usage)

If the generalizations have bugs, it is irrelevant when it doesn't affect the actual business-logic implementations.
If there's a bug in `shared` that's relevant, we'll notice it in the tests for the components above.

---

Aside: I used the term "cohesive generalizations" instead of "abstractions". What I mean is that a generalization is slightly different from an abstraction, but in the context of what I'm proposing, they should be treated just the same.
I'm not sure how much we should go into detail when publishing docs. To me, the devil is in the details, but others might not care whatsoever